### PR TITLE
Changes for web3.js migration

### PR DESCRIPTION
**Pull request by sjchmiela**
- [Loom (3 min)](https://www.loom.com/share/c38626e3bca54dc7a6750dc053e1ce7c)

Solana Pay Transaction Requests spec mentioning setting `1111` as fee payer:
Zrzut ekranu 2022-09-15 o 16 05 32

### CODE REVIEW

File: packages/solana-client/src/GTransaction.ts [hunk 1]
```diff
@@ -359,6 +359,72 @@ export namespace GTransaction {
...
+  export const updateFeePayer = ({
+    gtransaction,
+    feePayer,
+  }: {
+    gtransaction: GTransaction;
+    feePayer: Solana.Address;
+  }): GTransaction => {
+    const accountsMap = keyBy(gtransaction.accounts, "address");
+    // Recreate transaction not to copy old feePayer
+    const unsignedTransaction = create({
```
Comment (vpontis) [inline packages/solana-client/src/GTransaction.ts:372]: It would be nice if we had an easier way of copying txs but let's not change that in this PR.

**Review by vpontis — APPROVED** [review]

**Review by vpontis — CHANGES_REQUESTED** [review]
I'm a bit confused re `null` signatures. Why do we need to allow that?

**Comment by sjchmiela** [comment]
Mainly for coherence. Right now on `master` for a given snippet:

```ts
const original = GTransaction.create({ ..., signers: [] }); // { signatures: [{ address: "sth", signature: null }] }
const originalSerialized = GTransaction.toBuffer({ gtransaction: original });
const originalDeserialized = GTransaction.parse({ buffer: originalSerialized }); // { signatures: [{ address: "sth", signature: "1111" }] }
```

signatures are not equal anymore while the only thing we did was serialize and deserialize.

Another reason I mentioned is "compatibility" with `@solana/web3.js` API where they do exactly that — [parse `1111` signatures as `null`](https://github.com/solana-labs/solana/blob/bac6ebcd8814f35ba23bb2690895b27f84da494e/web3.js/src/transaction/legacy.ts#L828-L831) and we depend on it (at least) [here](https://github.com/luma-team/luma/blob/43f860500f2a854b9ff960a75225122556772d22/server/src/routes/solana-pay.ts#L30).

**Comment by vpontis** [comment]
> Mainly for coherence. Right now on `master` for a given snippet:
> 
> ```ts
> const original = GTransaction.create({ ..., signers: [] }); // { signatures: [{ address: "sth", signature: null }] }
> const originalSerialized = GTransaction.toBuffer({ gtransaction: original });
> const originalDeserialized = GTransaction.parse({ buffer: originalSerialized }); // { signatures: [{ address: "sth", signature: "1111" }] }
> ```
> 
> signatures are not equal anymore while the only thing we did was serialize and deserialize.
> 
> Another reason I mentioned is "compatibility" with `@solana/web3.js` API where they do exactly that — [parse `1111` signatures as `null`](https://github.com/solana-labs/solana/blob/bac6ebcd8814f35ba23bb2690895b27f84da494e/web3.js/src/transaction/legacy.ts#L828-L831) and we depend on it (at least) [here](https://github.com/luma-team/luma/blob/43f860500f2a854b9ff960a75225122556772d22/server/src/routes/solana-pay.ts#L30).

Hmm, can we not include the `null` change in this PR? I'm not convinced we need it yet. I can play around w/ our codebase to see how we can work around it. 

I don't think we need full compatibility w/ web3.js unless I'm missing something.

**Comment by sjchmiela** [comment]
Sure. The workaround for this is to instead of checking `signature === null` do:

```
signature === null || bs58.decode(signature).every(byte => byte === 0)
```

which felt inconvenient to me.

I agree we don't need full compatibility with web3.js, which is why I mentioned it as second, extra argument. The main reasons I think this change is nice is convenience (shown above) and equality between original and reserialized transactions.

Will revert signatures commit!

**Review by vpontis — APPROVED** [review]

File: packages/solana-client/src/GTransaction.ts [hunk 2]
```diff
@@ -409,3 +475,6 @@ const constructMessageBase64 = ({
 
   return messageBuffer.toString("base64");
 };
+
+const isSignatureEmpty = (signature: Base58 | null) =>
+  !signature || bs58.decode(signature).every((byte) => byte === 0);
```
Comment (vpontis) [inline packages/solana-client/src/GTransaction.ts:480]: Hmm, I don't think we need to decode here. I think it'll just be all `1`s

**Pull request merged**